From f500868f7cbda7642ac566e8feb98bc54d052be0 Mon Sep 17 00:00:00 2001
From: Subbaraya Sundeep <sbhatta@marvell.com>
Date: Wed, 24 Mar 2021 23:05:23 +0530
Subject: [PATCH 1380/1921] octeontx2-pf: Delete old TC node first

When installing TC flower rule currently mailbox
mutex is acquired for installing flow and old TC node
is deleted in critical section. But function which
deletes old node also acquires mailbox mutex hence
causing double lock problem. Hence delete old node
first then install flow for new node.

Fixes: 4f463d2e38f5 ("octeontx2-pf: Add tc flower hardware offload on")
Change-Id: I77f777e3a8760f6d458563447d5083dc11b74052
Signed-off-by: Subbaraya Sundeep <sbhatta@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/kernel/linux/+/48868
Tested-by: sa_ip-sw-jenkins <sa_ip-sw-jenkins@marvell.com>
Reviewed-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
[WK: The original patch got from Marvell sdk11.21.09]
Signed-off-by: Wenlin Kang <wenlin.kang@windriver.com>
---
 .../net/ethernet/marvell/octeontx2/nic/otx2_tc.c   | 14 +++++++-------
 1 file changed, 7 insertions(+), 7 deletions(-)

diff --git a/drivers/net/ethernet/marvell/octeontx2/nic/otx2_tc.c b/drivers/net/ethernet/marvell/octeontx2/nic/otx2_tc.c
index 93514b2aeaa1..9044c05a2118 100644
--- a/drivers/net/ethernet/marvell/octeontx2/nic/otx2_tc.c
+++ b/drivers/net/ethernet/marvell/octeontx2/nic/otx2_tc.c
@@ -534,6 +534,13 @@ static int otx2_tc_add_flow(struct otx2_nic *nic,
 		return -ENOMEM;
 	new_node->cookie = tc_flow_cmd->cookie;
 
+	/* If a flow exists with the same cookie, delete it */
+	old_node = rhashtable_lookup_fast(&tc_info->flow_table,
+					  &tc_flow_cmd->cookie,
+					  tc_info->flow_ht_params);
+	if (old_node)
+		otx2_tc_del_flow(nic, tc_flow_cmd);
+
 	mutex_lock(&nic->mbox.lock);
 	req = otx2_mbox_alloc_msg_npc_install_flow(&nic->mbox);
 	if (!req) {
@@ -548,13 +555,6 @@ static int otx2_tc_add_flow(struct otx2_nic *nic,
 		return rc;
 	}
 
-	/* If a flow exists with the same cookie, delete it */
-	old_node = rhashtable_lookup_fast(&tc_info->flow_table,
-					  &tc_flow_cmd->cookie,
-					  tc_info->flow_ht_params);
-	if (old_node)
-		otx2_tc_del_flow(nic, tc_flow_cmd);
-
 	if (bitmap_full(tc_info->tc_entries_bitmap, nic->flow_cfg->tc_max_flows)) {
 		netdev_err(nic->netdev, "Not enough MCAM space to add the flow\n");
 		otx2_mbox_reset(&nic->mbox.mbox, 0);
-- 
2.31.1

